rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+message DatabaseObjectSpec {
+   string Protocol = 1 [(gogoproto.jsontag) = "protocol,omitempty"];
+   string ServiceName = 2 [(gogoproto.jsontag) = "service_name,omitempty"];
+   string ObjectKind = 3 [(gogoproto.jsontag) = "object_kind"];
+   string Database = 4 [(gogoproto.jsontag) = "database,omitempty"];
+   string Schema = 5 [(gogoproto.jsontag) = "schema,omitempty"];
+   string Name = 6 [(gogoproto.jsontag) = "name,omitempty"];
+   // extra attributes for matching
+   map<string, string> Attributes = 7 [(gogoproto.jsontag) = "attributes,omitempty"];
+}
+```
+
+#### Role extension: `spec.{allow,deny}.db_permissions` fields
+
+The role is extended with a `db_permissions` field, which consists of a list of
+permissions to be applied against particular database objects provided the
+permission properties (object kind, match labels) match the given object.
+
+```protobuf
+
+  // ...
+  // DatabasePermission specifies a set of permissions that will be granted
+  // to the database user when using automatic database user provisioning.
+  repeated DatabasePermission DatabasePermissions = 38 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "db_permissions,omitempty"
+  ];
+}
+
+// DatabasePermission specifies the database object permission for the user.
+message DatabasePermission {
+  // ObjectKind is the database object kind: table, schema, etc.
+  string ObjectKind = 1 [(gogoproto.jsontag) = "object_kind"];
+  // Permission is the string representation of the permission to be given, e.g. SELECT, INSERT, UPDATE, ...
+  string Permission = 2 [(gogoproto.jsontag) = "permission"];
+  // Match is a list of labels (key, value) to match against database object properties.
+  map<string, string> Match = 3 [(gogoproto.jsontag) = "match,omitempty"];
+}
+
+```
+
+### Permission Semantics
+
+The precise meaning of individual permissions is left to the database engine.
+The sole exception is the interaction between the `deny` and `allow` parts:
+denied permissions are removed, and the comparison is performed in a
+case-insensitive way after trimming the whitespace. As a special case, `*` is
+allowed as a permission in `deny` part of the role.
+
+For example, if the `allow` permission is `SELECT`, then it can be removed with
+`select`, `SELECT` or `*`.
+
+The engines should prohibit invalid permissions. This is a fatal error and
+should cause connection error.
+
+### Permission Lifecycle
+
+Permissions are applied after the user is automatically provisioned.
+Deprovisioning (deletion/deactivation) of the user should remove ALL permissions
+from the user, without the need to reference the particular user permissions.
+
+The permission synchronization is performed at least once, on connection time:
+
+- database schema is read,
+- import rules are applied,
+- effective permissions are calculated based on user roles,
+- database-side permissions are updated.
+
+Optional, additional syncs may happen as needed (e.g. when a schema change is
```

Comment:
I'd say "no", to avoid extra complexity.

Reply:
I've added a note regarding updated roles: while technically possible, the extra complexity is probably not worth it, yet it may be revisited in the future as needed.